LibJS+WebContent: Implement console.table

GasimGasimzada opened the pull request:
## Description

Implement `console.table` similar to how it is handled in other browsers. In this case, the Console creates a data structure and passed that structure to the Printer. In the printer side, unlike in other browsers, I decided to allow expanding arrays and objects in the table instead of just saying that the element is an array or object:

![image](https://github.com/user-attachments/assets/f2c6aed9-4c68-4c02-b6ea-1f6240e83994)

## Algorithm

Below is the detailed explanation of the algorithm:

There are two parts to this algorithm -- First one is about choosing the column that represents indices and the second one is assigning properties to columns.

**Setting the index column:**

The index column's name is represented as "(index)" (parentheses are part of the column name). 

- If the first argument (let's call it `tabularData`) is a array, we iterate over the array and assign it to "(index)" column. 

<details><summary>Example:</summary>
   
   ```ts
   console.table(["a", "b", "c"]);
   ```

   outputs

   | (index) | Value |
   |--|--|
   | 0 | "a" |
   | 1 | "b" |
   | 2 | "c |

</details>

- If the first argument is an object, we iterate over all the enumerable keys and assign the key to the "(index)" column. 


   <details><summary>Example:</summary>

   ```ts
   console.table({ a: "10", b: 20, c: 30 });
   ```

   outputs

   | (index) | Value |
   |--|--|
   | a | 10 |
   | b | 20 |
   | c | 30 |

  </details>

</details>

**Assigning properties to columns:**

As we iterate over each item of the object/array, we check the type of each item (array element or object property):

- If the item is an array, iterate over the array, create a column that's named as the array element index and assign the value at the index to the column value.

    <details><summary>Example:</summary>

    ```ts
    console.table([
      ["hello", "world"]
      ["a", "b", "c"]
    ]);
    ```

   outputs

   | (index) | 0 | 1 | 2 |
   |---|---|---|---|
   | 0 | "hello" | "world" | |
   | 1 | "a" | "b" | "c" |

   </details>

- If the item is an object, enumerate over the object propreties, create a column that's named as the property key and assign the value at the key to the the property value.

   <details><summary>Example:</summary>

    ```ts
    console.table([
      {firstName: "hello", lastName: "world"}
      {a: 10, b: 20, c: 30]
    ]);
    ```

   outputs

   | (index) | firstName | lastName | a | b | c |
   |---|---|---|---|---|---|
   | 0 | "hello" | "world" | | | |
   | 1 |  |  | 10 | 20 | 30 |

   </details>

- If item is neither array, nor object, assign the object to column named "Value".

   <details><summary>Example:</summary>

   ```ts
   console.table([10, 20, 30]);
   ```

   outputs

   | (index) | Value |
   |--|--|
   | 0 | 10 |
   | 1 | 20 |
   | 2 | 30 |

  </details>

- When the second argument (let's call it `properties`) is passed, we show "(index)", "Value", the the columns defined in properties.

   <details><summary>Example:</summary>

    ```ts
    console.table([
      {firstName: "hello", lastName: "world"}
      {a: 10, b: 20, c: 30],
      "hello"
    ], ["firstName", "lastName"]);
    ```

    outputs

    | (index) | firstName | b | Value |
    |--|--|--|--|
    | 0 | "hello" | | |
    | 1 | | 20 | |
    | 2 | | | "hello" |

   </details>

**Considerations:**

- The creation of columns based on the item type is only done one level. If there is an array or object after first level, we just print it without creating new columns. 

   <details><summary>Example:</summary>

   ```ts
   console.table([
     [{ a: 10 }, { b:  20 }]
     { c: [10, 20], d: [30, 40] }
   ])
   ```

  Outputs:

  | (index) | 0 | 1 | c | d |
  |--|--|--|--|--|
  | 0 | { a: 10 } | { b: 20 } | |
  | 1 | | | [10, 20] | [30, 40] |

  </details>


- The order of elements can affect the order the columns, including the "Value" column, because columns are populated as we iterate over rows. 

   <details><summary>Example:</summary>

    ```ts
    console.table([
      {firstName: "hello", lastName: "world"}
      {a: 10, b: 20, c: 30],
      "test"
    ]);
    ```

   outputs

   | (index) | firstName | lastName | a | b | c | Value |
   |---|---|---|---|---|---|--|
   | 0 | "hello" | "world" | | | |
   | 1 |  |  | 10 | 20 | 30 |
   | 2 | "hello" | "world" | | | | test |

   while

    ```ts
    console.table([
      {a: 10, b: 20, c: 30]
      "test",
      {firstName: "hello", lastName: "world"}
    ]);
    ```

   outputs

   | (index) | a | b | c | Value | firstName | lastName |
   |---|---|---|---|---|---|--|
   | 0 |  10 | 20 | 30 | | | |
   | 1 | | | | test | | |
   | 2 | | | | | "hello" | "world" |

   </details>


- "(index)" is always the first column
- "(index)" and "Value" columns are **NOT** affected by the `properties` argument

## Spec Steps

**Spec PR: https://github.com/whatwg/console/pull/237 (still in progress)**


## Changes

- Expose table from console object
- Add new Table log level
- Create a JS object that represents table rows and columns
- Print table as HTML using WebContentConsoleClient

circl-lastname commented:
Commit should be titled `LibJS+WebContent"

AtkinsSJ commented on Base/res/ladybird/inspector.css:
```diff
@@ -147,11 +147,11 @@ html {
 
-.tab-controls button + button {
+.tab-controls button+button {
     border-left: 1px solid var(--tab-button-border);
 }
 
-details > :not(:first-child) {
+details> :not(:first-child) {
```
Please avoid making unrelated formatting changes.

GasimGasimzada replied:
Sorry, I had autoformatter that changed it without me noticing. I'll revert this change.

GasimGasimzada replied:
Fixed it!

AtkinsSJ commented on Base/res/ladybird/inspector.css:
```diff
@@ -271,3 +271,58 @@ details > :not(:first-child) {
+}
+
+
+@media (prefers-color-scheme: light) {
+    .console-log-table tbody tr:nth-of-type(2n+1) {
+        background-color: rgb(229, 229, 229);
+    }
+
+    .console-log-table tbody tr:hover {
+        background-color: rgb(199, 198, 198);
+    }
+}
+
+@media (prefers-color-scheme: dark) {
+    .console-log-table tbody tr:nth-of-type(2n+1) {
+        background-color: rgb(57, 57, 57);
+    }
+
+    .console-log-table tbody tr:hover {
+        background-color: rgb(80, 79, 79)
+    }
+}
```
This file was recently changed to use a single `@media` rule which then sets color variables, please do that here too.

GasimGasimzada replied:
Moved all colors to variables.

AtkinsSJ reviewed (changes requested):
Thanks for doing this! A couple of general comments besides the CSS ones:

I saw you made a PR to the console spec, that's great! Can you please include the spec steps you added as comments in this code? Put a note that links to the PR and says it's not yet in the actual spec. That'll make it easier to follow what's going on. (And help to verify that your algorithm works as intended.)

Also, the way the `properties` parameter works doesn't match what I see in Firefox or Chrome:
![image](https://github.com/user-attachments/assets/567d7002-d164-4e33-8e06-b86f1e84bba4)
(Ladybird on left, Firefox on right. Chrome output was the same as Firefox.)

GasimGasimzada commented:
> Thanks for doing this! A couple of general comments besides the CSS ones:
> 
> I saw you made a PR to the console spec, that's great! Can you please include the spec steps you added as comments in this code? Put a note that links to the PR and says it's not yet in the actual spec. That'll make it easier to follow what's going on. (And help to verify that your algorithm works as intended.)
> 
> Also, the way the `properties` parameter works doesn't match what I see in Firefox or Chrome: ![image](https://private-user-images.githubusercontent.com/222642/357458864-567d7002-d164-4e33-8e06-b86f1e84bba4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1Njc3MjIsIm5iZiI6MTcyMzU2NzQyMiwicGF0aCI6Ii8yMjI2NDIvMzU3NDU4ODY0LTU2N2Q3MDAyLWQxNjQtNGUzMy04ZTA2LWI4NmYxZTg0YmJhNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxM1QxNjQzNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTdlNGEwM2UyZDcwNGM0ODVlZGJmODY5YmNhMzkxMjZjMGRkMDFmMzBjYmIxMjUwYjhiMDg0ZjliZDEyYzQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NG_lNkQPqyip1GyPnRNIy38WOUZDy28D_BIZEDJ648w) (Ladybird on left, Firefox on right. Chrome output was the same as Firefox.)

Thank you for noticing this. Let me see if I can get it working properly (I will need to update the spec one as well)

GasimGasimzada commented:
@AtkinsSJ You are passing two different values between browsers:

On the left screenshot, the first argument is an object while on the right screenshot, the first argument is an array that contains an object:

![357458864-567d7002-d164-4e33-8e06-b86f1e84bba41](https://github.com/user-attachments/assets/fb1ecff2-9026-4276-8ec4-27ab89b34f1b)

AtkinsSJ commented:
Ah whoops! I'm glad it's me and not your code. 😆

AtkinsSJ commented:
OK, I think I wasn't clear with what I meant.

If you look at the other console methods, our code has the steps from the spec quoted exactly, then the code that implements them. For example, our `Console::count()` code looks like:

```c++
    // 1. Let map be the associated count map.
    auto& map = m_counters;

    // 2. If map[label] exists, set map[label] to map[label] + 1.
    if (auto found = map.find(label); found != map.end()) {
        map.set(label, found->value + 1);
    }
    // 3. Otherwise, set map[label] to 1.
    else {
        map.set(label, 1);
    }

	// ...
```

We like doing this for a few reasons: It makes it clear that the code matches the spec; it's easier to spot errors; and if the spec changes in the future, it's easier to see what needs changing. (I think someone was working on an automated way to check this, at some point...) As an added bonus it helps us spot any problems in the spec itself.

If you could do the same with your draft spec's steps, with a note at the top to say it's based on changes that aren't in the official spec yet, that would be good.

Also, your commit title still says "JSLib" when it should say "LibJS". :sweat_smile:

GasimGasimzada commented:
> If you could do the same with your draft spec's steps, with a note at the top to say it's based on changes that aren't in the official spec yet, that would be good.

Makes sense. I'll add the steps in the code 😊 

> Also, your commit title still says "JSLib" when it should say "LibJS". 😅

I'll fix that too. I thought this was about the PR title, not commit message title.

GasimGasimzada commented:
@AtkinsSJ I fixed the commit message and also added steps of the algorithm, and also added "WIP" next to the `table` function's comment. Let me know what you think.

AtkinsSJ commented on Userland/Libraries/LibJS/Console.cpp:
```diff
@@ -140,6 +142,198 @@ ThrowCompletionOr<Value> Console::log()
+                auto index = TRY(PropertyKey::from_value(vm, key));
+                auto value = TRY(object.get(index));
+
+                // 4.1.1. Perform create table row with `value`, `key`, `finalColumns`, and `properties` that returns `row`
+                auto row = TRY(create_table_row(realm(), key, value, final_columns, visited_columns, properties));
```
It's slightly confusing that the order in your spec text is "`value`, `key`" but the function takes them as "`key`, `value`". Same for the other callers of `create_table_row()`.

GasimGasimzada replied:
Fixed it. I changed the spec a bit and missed updating this part.

AtkinsSJ commented on Userland/Libraries/LibJS/Console.cpp:
```diff
@@ -140,6 +142,198 @@ ThrowCompletionOr<Value> Console::log()
+            // 5.4. Perform `Printer("table", finalData)`
+            MarkedVector<Value> args(vm.heap());
+            args.append(Value(final_data));
+            return m_client->logger(LogLevel::Table, args);
```
This is calling `logger()` not `printer()`.

GasimGasimzada replied:
Done

AtkinsSJ commented on Userland/Libraries/LibJS/Console.cpp:
```diff
@@ -140,6 +142,198 @@ ThrowCompletionOr<Value> Console::log()
+    }
+
+    // 6. Otherwise, perform `Printer("log", tabularData)`
+    return m_client->logger(LogLevel::Log, vm_arguments());
```
This is also calling `logger()` not `printer()`.

GasimGasimzada replied:
Done

AtkinsSJ reviewed (changes requested):
Thank you, it's a lot easier to examine the code when it's right next to the spec algorithm. (Even if it's a WIP algorithm.)

Someone else who's more familiar with LibJS will need to give this a look too, but overall this looks fine, just a few small things.

GasimGasimzada commented on Userland/Libraries/LibJS/Console.cpp:
```diff
@@ -140,6 +142,198 @@ ThrowCompletionOr<Value> Console::log()
+    auto& vm = realm.vm();
+
+    auto add_column = [&](PropertyKey const& column_name) -> Optional<Completion> {
+        if (!visited_columns.contains(column_name)) {
+            visited_columns.set(column_name, true);
```
Added a note explaining that it is an optimization.

AtkinsSJ reviewed (approved):
Great! I'm happy to merge this now, and if we do find any issues we can fix them later. Hopefully this will help demonstrate that the algorithm you wrote is correct, too. :^)
